### PR TITLE
DX: Specify value type in iterables in tests

### PR DESCRIPTION
**Pull request by jrmajor**

### CODE REVIEW

File: tests/AutoReview/CiConfigurationTest.php [hunk 1]
```diff
@@ -52,9 +52,11 @@ public function testTestJobsRunOnEachPhp(): void
             self::generateMinorVersionsRange($supportedMinPhp, $supportedMaxPhp)
         );
 
+        static::assertTrue(\count($supportedVersions) > 0);
```
Comment (jrmajor) [inline tests/AutoReview/CiConfigurationTest.php:55]: Prevents PHPStan from reporting that `end($supportedVersions)` may return `false`.

File: tests/Fixer/Operator/BinaryOperatorSpacesFixerTest.php [hunk 2]
```diff
@@ -2364,7 +2368,6 @@ public function provideFixPhp74Cases(): array
                     $a = fn()    =>      null;
                     $b = fn()      =>  null;
                 ',
-                null,
```
Comment (jrmajor) [inline tests/Fixer/Operator/BinaryOperatorSpacesFixerTest.php:2367]: Too many arguments for `($expected, $input, $configuration)`.

File: tests/RuleSet/RuleSetsTest.php [hunk 3]
```diff
@@ -134,7 +134,7 @@ public function testSetDefinitionsItselfIsSorted(): void
     {
         $setDefinition = array_keys(RuleSets::getSetDefinitions());
         $sortedSetDefinition = $setDefinition;
-        $this->sort($sortedSetDefinition);
+        sort($sortedSetDefinition);
```
Comment (jrmajor) [inline tests/RuleSet/RuleSetsTest.php:137]: Simple `list<string>`, no need for recursive/key-dependent sorting.

File: tests/AutoReview/CiConfigurationTest.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (jrmajor) [reply tests/AutoReview/CiConfigurationTest.php:55]: PHPStan does not understand `assertNotEmpty` and thinks that `end($supportedVersions)` may be null.

File: tests/Test/AbstractIntegrationCaseFactory.php [hunk 4]
```diff
@@ -105,7 +105,7 @@ protected function determineConfig(SplFileInfo $file, ?string $config): array
     /**
      * Parses the '--REQUIREMENTS--' block of a '.test' file and determines requirements.
      *
-     * @param ?string $config
+     * @return array<string, int>|array{php: int}
```
[the remark this answers is not in the document]
Comment (jrmajor) [reply tests/Test/AbstractIntegrationCaseFactory.php:108]: That would require more validation in this method, can we leave that for another PR?

**Comment by SpacePossum** [comment]
Most of the test of the fixers have dataproviders in the form og `expected`,  `input` and sometimes `config`, does typing down all these providers in PHPDoc's as well help. I wonder because the consumer of these, i.e. the `test*` functions already have these typed as being the input.
It adds a lot of paperwork doing the typing twice (so `test` method signature and again on the dataprovider), but I'm not sure what the gain is? It also makes the tests file longer and adding new tests more cumbersome.

File: tests/Test/AbstractIntegrationCaseFactory.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (julienfalque) [reply tests/Test/AbstractIntegrationCaseFactory.php:108]: Sure.

**Comment by julienfalque** [comment]
There's not much gain for now except making the number of reported errors lower. But hopefully someday PHPStan will be able to automatically check dataproviders against test methods' signature. I tried working on that (https://github.com/phpstan/phpstan-phpunit/issues/70) but since I lost motivation.

Anyway, regarding the pattern of the test methods, I think we can definitely refactor this somehow. Maybe we can also harmonize the different methods. I mean, sometimes the methods accept the configuration as last parameter, sometimes first, … Also some fixer tests have only one method with all tests with various configurations, sometimes one method per configuration, sometimes it's something else. Personnaly, I would like to always have one single `testFix($expected, $input, $config)` method and optionnally its PHP specific version variants.

**Review by julienfalque — APPROVED** [review]

**Comment by julienfalque** [comment]
Thank you @SpacePossum.

**Comment by SpacePossum** [comment]
>There's not much gain for now except making the number of reported errors lower. 

Think my point was, these aren't really errors, as we already enforce the typing in the test method signatures.

These new SCA rules forces all contributors to add these additional duplicate type hints to prevent the SCA from tripping, this is not ideal, especially if there is no much again. Also the tests are way bigger and harder to read.

For me, this is getting frustrating already, as PR's look red while there is nothing wrong. Reviewers might ignore PR's that are flagged by the CI. It also makes people loose focus on the actual meaningful change in a PR's as both authors and reviewers now also have to look into this noise.

**Comment by julienfalque** [comment]
I kind of agree but on a day to day basis I don't think this will be much bothering. Also, refactoring the fixers tests should reduce that noise a lot.

**Comment by jrmajor** [comment]
> It adds a lot of paperwork doing the typing twice (so test method signature and again on the dataprovider)

@SpacePossum But we don't do that, right? We ignore the pattern below in `phpstan.neon`, so that we do not need to type the data provider.

```neon
ignoreErrors:
    - '#^Method PhpCsFixer\\Tests\\.+::provide.+Cases\(\) return type has no value type specified in iterable type (array|iterable)\.$#'
```

> But hopefully someday PHPStan will be able to automatically check dataproviders against test methods' signature. 

@julienfalque FYI, Psalm already does that.

**Pull request merged**